packages/next/src/server/create-deduped-by-callsite-server-error-loger.ts

```diff
@@ -0,0 +1,63 @@
+import * as React from 'react'
+
+const errorRef: { current: null | string } = { current: null }
+// We don't want to dedupe across requests.
+// The developer might've just attempted to fix the warning so we should warn again if it still happens.
+const flushCurrentErrorIfNew = React.cache(
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars -- cache key
+  (key: unknown) => {
+    try {
+      console.error(errorRef.current)
+    } finally {
+      errorRef.current = null
+    }
+  }
+)
+
+/**
+ * Creates a function that logs an error message that is deduped by the userland
+ * callsite.
+ * This requires a known number of frames between the library and the userland
+ * callsite.
+ *
+ * Do not use on the Client where sourcemaps and ignore listing might be enabled.
+ * Only use that for warnings need a fix independent of the callstack.
+ *
+ * @param getMessage
+ * @params libraryFrames The number of frames between the library and the userland callsite.
+ *         For proxies that usually means 0. For Object.defineProperty it's 1.
+ * @returns
+ */
+export function createDedupedByCallsiteServerErrorLoggerDev<Args extends any[]>(
+  getMessage: (...args: Args) => string,
+  libraryFrames: number
+) {
+  return function logDedupedError(...args: Args) {
+    const message = getMessage(...args)
+
+    if (process.env.NODE_ENV !== 'production') {
+      const callStackFrames = new Error().stack?.split('\n')
+      if (
+        callStackFrames === undefined ||
+        callStackFrames.length < 3 + libraryFrames
+      ) {
+        console.error(message)
+      } else {
+        // Error:
+        //   logDedupedError
+        //   1st library frame
+        //   ....
+        //   nth library frame
+        //   <userland callsite>
+        // TODO: This breaks if sourcemaps with ignore lists are enabled.
+        // TODO: Let callsites pass in a function that we use in Error.captureStackTrace
+        // instead of defining the number of library frames and then use the top frame.
+        const key = callStackFrames[2 + libraryFrames]
```

Comment:
This brittle but covered by tests. E.g. I originally accidentally used a library frame as a key i.e. only warned once which broke the existing tests in `test/e2e/app-dir/dynamic-io/dynamic-io.search.test.ts`.

Using a library frame as a key when you do want to dedupe would also break test explicitly asserting on deduping.